File: engine/src/main/java/com/terralite/engine/weather/WeatherState.java
package com.terralite.engine.weather;

import java.util.Objects;

public final class WeatherState {
    private WeatherType type;
    private double intensity;

    public WeatherState() {
        this(WeatherType.CLEAR, 0.0);
    }

    public WeatherState(WeatherType type, double intensity) {
        set(type, intensity);
    }

    public WeatherType type() {
        return type;
    }

    public double intensity() {
        return intensity;
    }

    public void set(WeatherType type, double intensity) {
        this.type = Objects.requireNonNull(type, "type");
        if (intensity < 0.0 || intensity > 1.0) {
            throw new IllegalArgumentException("Weather intensity must be between 0 and 1");
        }
        this.intensity = intensity;
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/weather/WeatherState.kt
package com.terralite.engine.weather

class WeatherState @JvmOverloads constructor(
    private var type: WeatherType = WeatherType.CLEAR,
    private var intensity: Double = 0.0
) {
    init {
        validateIntensity(intensity)
    }

    fun type(): WeatherType = type
    fun intensity(): Double = intensity

    fun set(type: WeatherType, intensity: Double) {
        validateIntensity(intensity)
        this.type = type
        this.intensity = intensity
    }

    private companion object {
        fun validateIntensity(intensity: Double) {
            require(intensity in 0.0..1.0) { "Weather intensity must be between 0 and 1" }
        }
    }
}
